### PR TITLE
[x/gamm][stableswap]: Fix and expand gamm inverse relationship tests & fix bug in `CalcInAmtGivenOut`

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
## What is the purpose of the change

This PR fixes our inverse relationship tests which were previously not being run on both Balancer and Stableswap.

It also fixes a bug in our stableswap `CalcInAmtGivenOut`, and a separate bug in how the inverse relationship tests handles our stableswap CFMM for imbalanced pools.

As a byproduct of these changes, this PR also does basic level testing for the rounding-related changes added in #2912

TODO: Expand this functionality for the multi asset solver/calcs

## Brief Changelog

- Fix bug in inverse relationship testing suite that caused none of the tests to run
- Update the tests and testing methodology to properly accommodate base and edge cases for both stableswap and balancer
- Fix bug in `CalcInAmtGivenOut` that caused specific small inputs to be negated due to a failure to acknowledge error bounds
- As a simpler/more elegant solution to prevent the previous point from accidentally happening again, make solver support negative inputs

## Testing and Verifying

- All changes in this PR are tested in `amm_test.go` as part of the inverse relationship test suite

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

File: x/gamm/pool-models/stableswap/amm.go [hunk 1]
```diff
@@ -159,15 +159,16 @@ var (
...
-		} else if yIn.GTE(yReserve) {
+		} else if yIn.Abs().GTE(yReserve) {
 			panic("cannot input more than pool reserves")
 		}
 		k := constantFunction(xReserve, yReserve)
 		yFinal := yReserve.Add(yIn)
 		xLowEst := osmomath.ZeroDec()
-		xHighEst := xReserve
+		// we set upper bound at 2 * xReserve to accommodate negative yIns
+		xHighEst := xReserve.Mul(osmomath.NewBigDec(2))
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/amm.go:171]: Flagging that we should update this to be more in accordance with spec's discussion of bounds

**Review by ValarDragon — APPROVED** [review]
